pkg/controller/vitesscluster/reconcile_vtadmin.go

```diff
@@ -0,0 +1,373 @@
+			ExtraVolumeMounts: vt.Spec.VtAdmin.ExtraVolumeMounts,
+			InitContainers:    vt.Spec.VtAdmin.InitContainers,
+			SidecarContainers: vt.Spec.VtAdmin.SidecarContainers,
+			Annotations:       vt.Spec.VtAdmin.Annotations,
+			ExtraLabels:       vt.Spec.VtAdmin.ExtraLabels,
+			Tolerations:       vt.Spec.VtAdmin.Tolerations,
+		})
+	}
+	return specs, nil
+}
+
+func (r *ReconcileVitessCluster) createDiscoverySecret(ctx context.Context, vt *planetscalev2.VitessCluster, cell *planetscalev2.VitessCellTemplate) (*planetscalev2.SecretSource, error) {
+	// Get the vtctld service
+	vtctldService := corev1.Service{}
+	err := r.client.Get(ctx, client.ObjectKey{
+		Namespace: vt.Namespace,
+		Name:      vtctld.ServiceName(vt.Name),
+	}, &vtctldService)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the IP address from the service. This is randomly assigned.
+	vtctldServiceIP := vtctldService.Spec.ClusterIP
+	// The web and grpc ports should be set to the default values planetscalev2.DefaultWebPort and planetscalev2.DefaultGrpcPort
+	// respectively, but since we have the service, we can just read them.
+	var vtctldServiceWebPort, vtctldServiceGrpcPort int32
+	for _, port := range vtctldService.Spec.Ports {
+		if port.Name == planetscalev2.DefaultWebPortName {
+			vtctldServiceWebPort = port.Port
+		}
+		if port.Name == planetscalev2.DefaultGrpcPortName {
+			vtctldServiceGrpcPort = port.Port
+		}
+	}
+
+	// Read the cell information
+	vtc := planetscalev2.VitessCell{}
+	err = r.client.Get(ctx, client.ObjectKey{
+		Namespace: vt.Namespace,
+		Name:      vitesscell.Name(vt.Name, cell.Name),
+	}, &vtc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the vtgate service from the cell
+	vtgateService := corev1.Service{}
+	err = r.client.Get(ctx, client.ObjectKey{
+		Namespace: vtc.Namespace,
+		Name:      vtgate.ServiceName(vt.Name, cell.Name),
+	}, &vtgateService)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the IP address from the service. This is randomly assigned.
+	vtgateServiceIP := vtgateService.Spec.ClusterIP
+	// The grpc port should be set to the default value planetscalev2.DefaultGrpcPort,
+	// but since we have the service, we can just read it.
+	var vtgateServiceGrpcPort int32
+	for _, port := range vtgateService.Spec.Ports {
+		if port.Name == planetscalev2.DefaultGrpcPortName {
+			vtgateServiceGrpcPort = port.Port
+		}
+	}
+
+	// Variables to hold the key, value and secret name to use
+	discoveryKey := "discovery.json"
+	discoveryVal := fmt.Sprintf(`{
+    "vtctlds": [
+        {
+            "host": {
+                "fqdn": "%s:%d",
+                "hostname": "%s:%d"
+            }
+        }
+    ],
+    "vtgates": [
+        {
+            "host": {
+                "hostname": "%s:%d"
+            }
+        }
+    ]
+}`, vtctldServiceIP, vtctldServiceWebPort, vtctldServiceIP, vtctldServiceGrpcPort, vtgateServiceIP, vtgateServiceGrpcPort)
+	secretName := vtadmin.DiscoverySecretName(vt.Name, cell.Name)
```

Comment:
The discovery configuration file is autogenerated using the global service of `vtctld` and per cell service of `vtgates`.